tfx/orchestration/pipeline.py

```diff
@@ -189,3 +190,76 @@ def components(self, components: List[base_node.BaseNode]):
     # has all its dependencies visited.
     if len(self._components) < len(deduped_components):
       raise RuntimeError('There is a cycle in the pipeline')
+
+  def get_fuseable_components(self):
+    """
+       Returns a list of fuseable Beam component subraphs in sorted topological
+       order. For example:
+
+       [[component_a, component_b],[component_c, component_d, component_e]]
+
+       Experimental, backward compatability not guaranteed.
+    """
+
+    # Finds all Beam based components
+    beam_components = set()
+
+    for component in self.components:
+      if issubclass(component.EXECUTOR_SPEC.executor_class,
+                    base_executor.FuseableBeamExecutor):
+        beam_components.add(component)
+
+    # Finds Beam component "sources" (components with no parents that are Beam
+    # based components).
+    beam_component_sources = []
+
+    for component in sorted(beam_components, key=lambda c: c.id):
+      is_source = True
+
+      for parent in component.upstream_nodes:
+        if parent in beam_components:
+          is_source = False
+          break
+
+      if is_source:
+        beam_component_sources.append(component)
+
+    # Finds subgraphs of fuseable beam components, sorted in topological order
+    fuseable_components = []
+    visited = set()
+
+    for source in beam_component_sources: # pylint: disable=too-many-nested-blocks
+      fuseable_component = []
+      current_layer = [source]
+
+      # Conducts BFS search to build out fuseable Beam component subgraph
```

Comment:
I think a more complete high level description of this algorithm is needed here, because it's not a simple BFS: you're performing multiple BFS traversals, but you're maintaining a set `visited` between the traversals. I'm not entirely sure that the algorithm is optimal or even correct, or if it can be simplified.

For this PR, let's get some high level documentation and a correct algorithm, and worry about simplicity and optimality later.

This is more complicated than it looks and I had some trouble formulating the problem accurately. I think it is something like:

Given a graph with vertices V and edges E, and a set of marked edges F where F ⊆ E, output a partitioning (set of disjoint subsets) of vertices in V (e.g. {{V<sub>1</sub>, V<sub>2</sub>, V<sub>3</sub>}, {V<sub>4</sub>}, ...} where:
- Each partition is connected.
- The internal edges of each partition are a subset of F.

And I think the optimality condition is that no two partitions can be merged without violating one of the two conditions above. But I'm not entirely sure if the optimality condition is correct.

Reply:
High level documentation and the correct algorithm can be found [here](https://docs.google.com/document/d/1RGrAgqBuKb8n_z77lerBvVofUOoKNHW_cI-DjfFBAx8/edit).